File: monitor-daemon/collector/fps.go
package collector

import (
	"bufio"
	"context"
	"fmt"
	"math"
	"os/exec"
	"strings"
	"sync"
	"time"
)

// cmdTimeout is the maximum time allowed for any dumpsys command.
// If a command takes longer, it is killed to prevent the sampling goroutine from hanging.
const cmdTimeout = 5 * time.Second

// FpsResult is the current FPS snapshot exposed to callers.
// Pointer fields are nil before the first successful collection (JSON null).
type FpsResult struct {
	FPS     *float64 `json:"fps"`
	Jank    *int     `json:"jank"`
	BigJank *int     `json:"big_jank"`
	Layer   string   `json:"layer"`
	Source  string   `json:"source"`
}

// jankState holds per-layer PerfDog jank detection state.
// Reset on every app switch so counters always reflect the current app session.
type jankState struct {
	prevTs           int64   // last processed actualPresentTime (ns since boot)
	prevFrameTimeMs  float64 // interval of previous frame (ms)
	prev2FrameTimeMs float64 // interval of the frame before that (ms)
}

// fpsCollector samples SurfaceFlinger frame counts in the background.
// No smoothing — raw FPS values are output directly for instant transitions.
type fpsCollector struct {
	mu     sync.RWMutex
	result FpsResult

	prevCounts   map[string]int64
	prevTime     time.Time
	prevFocusPkg string

	trackedLayer string

	// PerfDog-style per-frame jank detection via --latency timestamps.
	jank jankState

	zeroStreak  int
	lastFrameNs int64 // latest actualPresentTime from --latency (ns since boot)

	// Recovery: consecutive sample failures trigger timestats re-enable.
	consecutiveErrors int
	lastReenableTime  time.Time
}

func newFpsCollector() *fpsCollector {
	return &fpsCollector{
		prevCounts: make(map[string]int64),
	}
}

func (fc *fpsCollector) start() {
	go func() {
		fc.enableTimestats()
		logInfo("fps", "timestats enabled, sampling every 500ms")

		for {
			fc.sample()
			time.Sleep(500 * time.Millisecond)
		}
	}()
}

// enableTimestats resets the SurfaceFlinger timestats state.
func (fc *fpsCollector) enableTimestats() {
	ctx, cancel := context.WithTimeout(context.Background(), cmdTimeout)
	defer cancel()
	exec.CommandContext(ctx, "dumpsys", "SurfaceFlinger", "--timestats", "-enable").Run() //nolint

	ctx2, cancel2 := context.WithTimeout(context.Background(), cmdTimeout)
	defer cancel2()
	exec.CommandContext(ctx2, "dumpsys", "SurfaceFlinger", "--timestats", "-clear").Run() //nolint

	fc.lastReenableTime = time.Now()
}

// reenableIfNeeded re-enables timestats after consecutive errors or long runtime.
// SurfaceFlinger may internally disable timestats or become unresponsive.
func (fc *fpsCollector) reenableIfNeeded() {
	if fc.consecutiveErrors >= 6 || time.Since(fc.lastReenableTime) > 10*time.Minute {
		logInfo("fps", "re-enabling timestats (consecutive_errors=%d, last_reenable=%s ago)",
			fc.consecutiveErrors, time.Since(fc.lastReenableTime).Round(time.Second))
		fc.enableTimestats()
		fc.consecutiveErrors = 0
		// Reset state so we rebuild baselines
		fc.prevCounts = make(map[string]int64)
		fc.prevTime = time.Time{}
	}
}

func (fc *fpsCollector) get() FpsResult {
	fc.mu.RLock()
	defer fc.mu.RUnlock()
	return fc.result
}

// ─── Foreground detection ───────────────────────────────────────────

func getFocusedPkg() string {
	if pkg := getFocusedFromActivity(); pkg != "" {
		return pkg
	}
	return getFocusedFromWindow()
}

func getFocusedFromActivity() string {
	ctx, cancel := context.WithTimeout(context.Background(), cmdTimeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, "dumpsys", "activity", "activities").Output()
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(string(out), "\n") {
		if strings.Contains(line, "mResumedActivity") {
			return extractPkgFromRecord(line)
		}
	}
	return ""
}

func getFocusedFromWindow() string {
	ctx, cancel := context.WithTimeout(context.Background(), cmdTimeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, "dumpsys", "window", "displays").Output()
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(string(out), "\n") {
		if strings.Contains(line, "mFocusedApp=") {
			return extractPkgFromRecord(line)
		}
	}
	return ""
}

func extractPkgFromRecord(line string) string {
	for _, prefix := range []string{" u0 ", " u1 ", " u2 ", " u10 "} {
		idx := strings.Index(line, prefix)
		if idx >= 0 {
			rest := line[idx+len(prefix):]
			slash := strings.Index(rest, "/")
			if slash > 0 {
				pkg := rest[:slash]
				if strings.Contains(pkg, ".") {
					return pkg
				}
			}
		}
	}
	return ""
}

// ─── Core sampling logic ────────────────────────────────────────────

func (fc *fpsCollector) sample() {
	fc.reenableIfNeeded()

	pkg := getFocusedPkg()
	if pkg == "" {
		fc.consecutiveErrors++
		return
	}

	counts, err := parseTimestats(pkg)
	if err != nil {
		fc.consecutiveErrors++
		return
	}
	fc.consecutiveErrors = 0

	now := time.Now()

	fc.mu.Lock()
	defer fc.mu.Unlock()

	// No matching layers for foreground package — count as idle
	if len(counts) == 0 {
		fc.zeroStreak++
		if fc.zeroStreak >= 2 {
			fc.result.FPS = float64Ptr(0)
		}
		return
	}

	// ── App switch: full reset ──
	if pkg != fc.prevFocusPkg {
		logInfo("fps", "app switch: %s → %s", fc.prevFocusPkg, pkg)
		fc.prevFocusPkg = pkg
		fc.prevCounts = make(map[string]int64)
		fc.trackedLayer = ""
		fc.prevTime = time.Time{}
		fc.result.FPS = float64Ptr(0)
		fc.result.Jank = intPtr(0)
		fc.result.BigJank = intPtr(0)
		fc.zeroStreak = 0
		fc.lastFrameNs = 0
		fc.jank = jankState{}
		for k, v := range counts {
			fc.prevCounts[k] = v
		}
		fc.prevTime = now
		return
	}

	if fc.prevTime.IsZero() {
		for k, v := range counts {
			fc.prevCounts[k] = v
		}
		fc.prevTime = now
		return
	}

	dt := now.Sub(fc.prevTime).Seconds()
	fc.prevTime = now
	if dt <= 0 {
		return
	}

	// ── Compute delta for ALL layers ──
	deltas := make(map[string]int64, len(counts))
	for layer, curCount := range counts {
		prevCount := fc.prevCounts[layer]
		delta := curCount - prevCount
		if delta > 0 {
			deltas[layer] = delta
		}
		fc.prevCounts[layer] = curCount
	}

	bestLayer, bestDelta := fc.selectLayer(deltas)

	if bestDelta <= 0 {
		// Timestats delta=0 — probe --latency to distinguish frozen counter from idle.
		// Idle layers return 0 timestamps; active layers (even during rate change) have data.
		if fc.probeLatency() {
			return
		}
		fc.zeroStreak++
		if fc.zeroStreak >= 2 {
			fc.result.FPS = float64Ptr(0)
		}
		return
	}

	fc.zeroStreak = 0
	fc.lastFrameNs = 0 // reset latency baseline when timestats is working

	if bestLayer != fc.trackedLayer && fc.trackedLayer != "" {
		logInfo("fps", "layer switch: %s → %s (delta=%d)", fc.trackedLayer, bestLayer, bestDelta)
	}
	fc.trackedLayer = bestLayer

	rawFps := float64(bestDelta) / dt

	// ── No smoothing — direct output ──
	fc.result.FPS = float64Ptr(math.Round(rawFps*10) / 10)
	fc.result.Layer = bestLayer
	fc.result.Source = "timestats"

	// ── PerfDog jank detection from per-frame --latency timestamps ──
	fc.detectJankPerfdog()
}

// ─── Sticky layer selection ──────────────────────────────────────────

func (fc *fpsCollector) selectLayer(deltas map[string]int64) (string, int64) {
	if len(deltas) == 0 {
		return fc.trackedLayer, 0
	}

	trackedDelta := deltas[fc.trackedLayer]

	if trackedDelta > 0 {
		for layer, delta := range deltas {
			if layer != fc.trackedLayer && delta > trackedDelta*2 {
				return layer, delta
			}
		}
		return fc.trackedLayer, trackedDelta
	}

	var bestLayer string
	var bestDelta int64
	bestIsBLAST := false

	for layer, delta := range deltas {
		isBLAST := isBLASTLayer(layer)
		switch {
		case delta > bestDelta:
			bestLayer = layer
			bestDelta = delta
			bestIsBLAST = isBLAST
		case delta == bestDelta && isBLAST && !bestIsBLAST:
			bestLayer = layer
			bestIsBLAST = true
		}
	}
	return bestLayer, bestDelta
}

func isBLASTLayer(layerName string) bool {
	return strings.Contains(layerName, "SurfaceView") && strings.Contains(layerName, "BLAST")
}

// ─── Latency probe (fallback when timestats delta=0) ────────────────

// probeLatency checks --latency for actual frame data.
// Returns true if frames were found (FPS updated or baseline set).
func (fc *fpsCollector) probeLatency() bool {
	if fc.trackedLayer == "" {
		return false
	}
	layerName := strings.TrimPrefix(fc.trackedLayer, "layerName = ")
	timestamps := parseLatency(layerName)
	if len(timestamps) == 0 {
		return false // no frame data → genuinely idle
	}

	// Find latest timestamp and count frames newer than baseline
	var latest int64
	var newCount int
	for _, ts := range timestamps {
		if ts > latest {
			latest = ts
		}
		if fc.lastFrameNs > 0 && ts > fc.lastFrameNs {
			newCount++
		}
	}

	// First probe: establish baseline, hold previous FPS
	if fc.lastFrameNs == 0 {
		fc.lastFrameNs = latest
		return true
	}

	if newCount == 0 {
		return false // all timestamps are old → idle
	}

	dt := float64(latest-fc.lastFrameNs) / 1e9
	fc.lastFrameNs = latest

	if dt >= 0.01 {
		fc.result.FPS = float64Ptr(math.Round(float64(newCount) / dt * 10) / 10)
		fc.result.Layer = fc.trackedLayer
		fc.result.Source = "latency"
		fc.zeroStreak = 0
		logInfo("fps", "latency: %d new frames in %.3fs → %.1f fps", newCount, dt, fc.result.FPS)
		return true
	}
	return false
}

// parseLatency runs `dumpsys SurfaceFlinger --latency <layer>` and returns
// valid actualPresentTime values (nanoseconds since boot).
func parseLatency(layer string) []int64 {
	ctx, cancel := context.WithTimeout(context.Background(), cmdTimeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, "dumpsys", "SurfaceFlinger", "--latency", layer).Output()
	if err != nil {
		return nil
	}
	lines := strings.Split(strings.TrimSpace(string(out)), "\n")
	if len(lines) < 2 {
		return nil
	}
	var timestamps []int64
	for _, line := range lines[1:] { // skip vsync period line
		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}
		ts, err := parseInt64(fields[1]) // actualPresentTime
		if err != nil || ts <= 0 || ts > 1e18 {
			continue
		}
		timestamps = append(timestamps, ts)
	}
	return timestamps
}

// ─── PerfDog jank detection ─────────────────────────────────────────
//
// Reads actualPresentTime stamps from `dumpsys SurfaceFlinger --latency`
// and applies the PerfDog algorithm per frame:
//
//   Jank    : frameTime > 2 × avg(prev2) AND frameTime > 2 × baseline
//   BigJank : frameTime > 3 × avg(prev2) AND frameTime > 3 × baseline
//
// baseline = 1000ms / 60fps = 16.67ms  (conservative; works on any refresh rate)
//
// Only frames with actualPresentTime newer than the last processed timestamp
// are evaluated, so each frame is counted exactly once across calls.

const jankBaselineMs = 1000.0 / 60.0 // 16.67 ms — one frame at 60 Hz

func (fc *fpsCollector) detectJankPerfdog() {
	if fc.trackedLayer == "" {
		return
	}
	layerName := strings.TrimPrefix(fc.trackedLayer, "layerName = ")
	timestamps := parseLatency(layerName)
	if len(timestamps) < 2 {
		return
	}

	for i := 1; i < len(timestamps); i++ {
		ts := timestamps[i]
		if ts <= fc.jank.prevTs {
			continue // already processed in a previous call
		}

		prev := timestamps[i-1]
		if prev <= 0 {
			fc.jank.prevTs = ts
			continue
		}

		frameTimeMs := float64(ts-prev) / 1e6
		if frameTimeMs <= 0 || frameTimeMs > 500 { // sanity: ignore gaps > 500ms (pause/idle)
			fc.jank.prevTs = ts
			continue
		}

		// Need two previous frames to compute the rolling average.
		if fc.jank.prevFrameTimeMs > 0 && fc.jank.prev2FrameTimeMs > 0 {
			avg := (fc.jank.prevFrameTimeMs + fc.jank.prev2FrameTimeMs) / 2.0
			if frameTimeMs > 3*avg && frameTimeMs > 3*jankBaselineMs {
				if fc.result.BigJank != nil {
					fc.result.BigJank = intPtr(*fc.result.BigJank + 1)
				}
			} else if frameTimeMs > 2*avg && frameTimeMs > 2*jankBaselineMs {
				if fc.result.Jank != nil {
					fc.result.Jank = intPtr(*fc.result.Jank + 1)
				}
			}
		}

		fc.jank.prev2FrameTimeMs = fc.jank.prevFrameTimeMs
		fc.jank.prevFrameTimeMs = frameTimeMs
		fc.jank.prevTs = ts
	}
}

// ─── Timestats parsing ──────────────────────────────────────────────

func parseTimestats(pkg string) (map[string]int64, error) {
	ctx, cancel := context.WithTimeout(context.Background(), cmdTimeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, "dumpsys", "SurfaceFlinger", "--timestats", "-dump").Output()
	if err != nil {
		return nil, err
	}

	result := make(map[string]int64)
	seen := make(map[string]bool)
	var inLayer bool
	var curLayer string

	sc := bufio.NewScanner(strings.NewReader(string(out)))
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())

		if strings.HasPrefix(line, "layerName =") {
			if strings.Contains(line, pkg) {
				if !seen[line] {
					seen[line] = true
					inLayer = true
					curLayer = line
				} else {
					inLayer = false
				}
			} else {
				inLayer = false
			}
			continue
		}

		if inLayer && strings.HasPrefix(line, "totalFrames =") {
			var count int64
			_, err := parseIntField(line, "totalFrames =", &count)
			if err == nil {
				result[curLayer] = count
			}
			inLayer = false
		}
	}
	return result, nil
}

func parseIntField(line, prefix string, out *int64) (string, error) {
	val := strings.TrimPrefix(line, prefix)
	val = strings.TrimSpace(val)
	n, err := parseInt64(val)
	if err == nil {
		*out = n
	}
	return val, err
}

func parseInt64(s string) (int64, error) {
	s = strings.TrimSpace(s)
	var n int64
	_, err := fmt.Sscanf(s, "%d", &n)
	return n, err
}

File: server-rs/src/collectors/fps.rs
//! FPS collector — ports the Go daemon's `collector/fps.go` to Rust.
//!
//! ARCHITECTURE:
//! - Foreground package: driven by ITaskStackListener event (task_stack.rs).
//!   Each sample copies the current value from a shared Mutex<String>.
//! - FPS / jank / timestats: binder `dump()` on SurfaceFlinger with
//!   `["--timestats", "-dump"]`. Zero fork, zero shell.
//! - Jank detection: PerfDog algorithm on per-frame actualPresentTime
//!   from `["--latency", layer]`.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use crate::events::sf_dump;

/// One frame at 60 Hz = 16.67 ms. Used as jank baseline.
const JANK_BASELINE_MS: f64 = 1000.0 / 60.0;

/// Max idle gap in ms to ignore in jank detection (pause/suspend).
const JANK_MAX_GAP_MS: f64 = 500.0;

/// Consecutive failures before re-enabling timestats.
const REENABLE_FAILURE_COUNT: u32 = 6;

/// Minimum interval between timestats re-enable attempts.
const REENABLE_INTERVAL_SECS: u64 = 600; // 10 minutes

pub struct FpsSample {
    pub fps_x100: i32,
    pub jank: i32,
    pub big_jank: i32,
    pub layer: String,
    pub focused_pkg: String,
}

pub struct FpsReader {
    foreground_pkg: Arc<Mutex<String>>,

    prev_counts: HashMap<String, i64>,
    prev_time: Option<Instant>,
    prev_focus_pkg: String,
    tracked_layer: String,
    zero_streak: u32,
    last_frame_ns: i64,
    jank_state: JankState,

    // Recovery
    consecutive_errors: u32,
    last_enable_time: Option<Instant>,

    // Accumulated jank counts per sample window
    jank_count: i32,
    big_jank_count: i32,
}

#[derive(Default)]
struct JankState {
    prev_ts: i64,
    prev_frame_time_ms: f64,
    prev2_frame_time_ms: f64,
}

impl FpsReader {
    pub fn open() -> Self {
        // Enable timestats on startup
        sf_dump::enable_timestats();
        Self {
            foreground_pkg: Arc::new(Mutex::new(String::new())),
            prev_counts: HashMap::new(),
            prev_time: None,
            prev_focus_pkg: String::new(),
            tracked_layer: String::new(),
            zero_streak: 0,
            last_frame_ns: 0,
            jank_state: JankState::default(),
            consecutive_errors: 0,
            last_enable_time: Some(Instant::now()),
            jank_count: 0,
            big_jank_count: 0,
        }
    }

    /// Returns a shared handle to the foreground package.
    /// Called by `events::task_stack` to share the listener's state.
    pub fn attach_focus(&self) -> Arc<Mutex<String>> {
        self.foreground_pkg.clone()
    }

    pub fn sample(&mut self) -> FpsSample {
        self.reenable_if_needed();

        let focused = self
            .foreground_pkg
            .lock()
            .map(|g| g.clone())
            .unwrap_or_default();

        if focused.is_empty() {
            self.consecutive_errors += 1;
            return self.make_sample(focused);
        }

        let text = match sf_dump::dump_timestats() {
            Some(t) => t,
            None => {
                self.consecutive_errors += 1;
                return self.make_sample(focused);
            }
        };

        let counts = sf_dump::parse_timestats(&text, &focused);
        self.consecutive_errors = 0;

        let now = Instant::now();

        // No matching layers for foreground package
        if counts.is_empty() {
            self.zero_streak += 1;
            if self.zero_streak >= 2 {
                return self.make_sample_with_fps(0, focused);
            }
            return self.make_sample(focused);
        }

        // App switch: full reset
        if focused != self.prev_focus_pkg {
            log::info!("fps: app switch: {} → {}", self.prev_focus_pkg, focused);
            self.prev_focus_pkg = focused.clone();
            self.prev_counts = counts;
            self.tracked_layer.clear();
            self.prev_time = Some(now);
            self.zero_streak = 0;
            self.last_frame_ns = 0;
            self.jank_state = JankState::default();
            self.jank_count = 0;
            self.big_jank_count = 0;
            return self.make_sample_with_fps(0, focused);
        }

        let prev_time = match self.prev_time {
            Some(t) => t,
            None => {
                self.prev_counts = counts;
                self.prev_time = Some(now);
                return self.make_sample(focused);
            }
        };

        let dt = now.duration_since(prev_time).as_secs_f64();
        self.prev_time = Some(now);
        if dt <= 0.0 {
            return self.make_sample(focused);
        }

        // Compute deltas for all layers
        let mut deltas = HashMap::new();
        for (layer, cur_count) in &counts {
            let prev = self.prev_counts.get(layer).copied().unwrap_or(0);
            let delta = cur_count - prev;
            if delta > 0 {
                deltas.insert(layer.clone(), delta);
            }
        }
        self.prev_counts = counts;

        let (best_layer, best_delta) = self.select_layer(&deltas);

        if best_delta <= 0 {
            // Probe latency as fallback
            if self.probe_latency(&focused) {
                return self.make_sample(focused);
            }
            self.zero_streak += 1;
            if self.zero_streak >= 2 {
                return self.make_sample_with_fps(0, focused);
            }
            return self.make_sample(focused);
        }

        self.zero_streak = 0;
        self.last_frame_ns = 0;

        if !best_layer.is_empty() && best_layer != self.tracked_layer && !self.tracked_layer.is_empty() {
            log::info!(
                "fps: layer switch: {} → {} (delta={})",
                self.tracked_layer,
                best_layer,
                best_delta
            );
        }
        self.tracked_layer = best_layer;

        let raw_fps = best_delta as f64 / dt;
        let fps_x100 = (raw_fps * 100.0).round() as i32;

        // Jank detection from latency timestamps
        self.detect_jank_perfdog();

        let mut s = self.make_sample_with_fps(fps_x100, focused);
        s.layer = self.tracked_layer.clone();
        s.jank = self.jank_count;
        s.big_jank = self.big_jank_count;
        // Reset per-sample jank counters
        self.jank_count = 0;
        self.big_jank_count = 0;
        s
    }

    fn make_sample(&self, focused_pkg: String) -> FpsSample {
        FpsSample {
            fps_x100: -1,
            jank: 0,
            big_jank: 0,
            layer: self.tracked_layer.clone(),
            focused_pkg,
        }
    }

    fn make_sample_with_fps(&self, fps_x100: i32, focused_pkg: String) -> FpsSample {
        FpsSample {
            fps_x100,
            jank: 0,
            big_jank: 0,
            layer: self.tracked_layer.clone(),
            focused_pkg,
        }
    }

    /// Sticky layer selection with 2× threshold for switching.
    /// Prefers BLAST layers on tie.
    fn select_layer(&self, deltas: &HashMap<String, i64>) -> (String, i64) {
        if deltas.is_empty() {
            return (self.tracked_layer.clone(), 0);
        }

        let tracked_delta = deltas.get(&self.tracked_layer).copied().unwrap_or(0);

        if tracked_delta > 0 {
            for (layer, &delta) in deltas {
                if layer != &self.tracked_layer && delta > tracked_delta * 2 {
                    return (layer.clone(), delta);
                }
            }
            return (self.tracked_layer.clone(), tracked_delta);
        }

        let mut best_layer = String::new();
        let mut best_delta: i64 = 0;
        let mut best_is_blast = false;

        for (layer, &delta) in deltas {
            let is_blast = is_blast_layer(layer);
            if delta > best_delta || (delta == best_delta && is_blast && !best_is_blast) {
                best_layer = layer.clone();
                best_delta = delta;
                best_is_blast = is_blast;
            }
        }
        (best_layer, best_delta)
    }

    /// Probe `--latency` when timestats delta is 0. Returns true if frames found.
    fn probe_latency(&mut self, _focused: &str) -> bool {
        if self.tracked_layer.is_empty() {
            return false;
        }
        let layer_name = self
            .tracked_layer
            .strip_prefix("layerName = ")
            .unwrap_or(&self.tracked_layer);

        let text = match sf_dump::dump_latency(layer_name) {
            Some(t) => t,
            None => return false,
        };
        let timestamps = sf_dump::parse_latency(&text);
        if timestamps.is_empty() {
            return false;
        }

        let latest = *timestamps.iter().max().unwrap_or(&0);

        if self.last_frame_ns == 0 {
            self.last_frame_ns = latest;
            return true;
        }

        let new_count = timestamps
            .iter()
            .filter(|&&ts| ts > self.last_frame_ns)
            .count();

        if new_count == 0 {
            return false;
        }

        let dt = (latest - self.last_frame_ns) as f64 / 1e9;
        self.last_frame_ns = latest;

        if dt >= 0.01 {
            let _fps = new_count as f64 / dt;
            self.zero_streak = 0;
            return true;
        }
        false
    }

    /// PerfDog-style jank detection from per-frame actualPresentTime.
    fn detect_jank_perfdog(&mut self) {
        if self.tracked_layer.is_empty() {
            return;
        }
        let layer_name = self
            .tracked_layer
            .strip_prefix("layerName = ")
            .unwrap_or(&self.tracked_layer);

        let text = match sf_dump::dump_latency(layer_name) {
            Some(t) => t,
            None => return,
        };
        let timestamps = sf_dump::parse_latency(&text);
        if timestamps.len() < 2 {
            return;
        }

        for i in 1..timestamps.len() {
            let ts = timestamps[i];
            if ts <= self.jank_state.prev_ts {
                continue;
            }

            let prev = timestamps[i - 1];
            if prev <= 0 {
                self.jank_state.prev_ts = ts;
                continue;
            }

            let frame_time_ms = (ts - prev) as f64 / 1e6;
            if frame_time_ms <= 0.0 || frame_time_ms > JANK_MAX_GAP_MS {
                self.jank_state.prev_ts = ts;
                continue;
            }

            if self.jank_state.prev_frame_time_ms > 0.0
                && self.jank_state.prev2_frame_time_ms > 0.0
            {
                let avg = (self.jank_state.prev_frame_time_ms
                    + self.jank_state.prev2_frame_time_ms)
                    / 2.0;
                if frame_time_ms > 3.0 * avg && frame_time_ms > 3.0 * JANK_BASELINE_MS {
                    self.big_jank_count += 1;
                } else if frame_time_ms > 2.0 * avg && frame_time_ms > 2.0 * JANK_BASELINE_MS {
                    self.jank_count += 1;
                }
            }

            self.jank_state.prev2_frame_time_ms = self.jank_state.prev_frame_time_ms;
            self.jank_state.prev_frame_time_ms = frame_time_ms;
            self.jank_state.prev_ts = ts;
        }
    }

    /// Re-enable timestats after consecutive failures or timeout.
    fn reenable_if_needed(&mut self) {
        let should_reenable = self.consecutive_errors >= REENABLE_FAILURE_COUNT
            || self
                .last_enable_time
                .map(|t| t.elapsed().as_secs() >= REENABLE_INTERVAL_SECS)
                .unwrap_or(true);

        if should_reenable {
            log::info!("fps: re-enabling timestats (errors={}, age={:?})",
                self.consecutive_errors,
                self.last_enable_time.map(|t| t.elapsed()));
            sf_dump::enable_timestats();
            self.consecutive_errors = 0;
            self.last_enable_time = Some(Instant::now());
        }
    }
}

fn is_blast_layer(name: &str) -> bool {
    name.contains("SurfaceView") && name.contains("BLAST")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn select_layer_prefers_tracked() {
        let reader = FpsReader {
            tracked_layer: "layerName = A".into(),
            ..make_test_reader()
        };
        let mut deltas = HashMap::new();
        deltas.insert("layerName = A".into(), 100);
        deltas.insert("layerName = B".into(), 150); // not 2× of 100
        let (layer, delta) = reader.select_layer(&deltas);
        assert_eq!(layer, "layerName = A");
        assert_eq!(delta, 100);
    }

    #[test]
    fn select_layer_switches_at_2x() {
        let reader = FpsReader {
            tracked_layer: "layerName = A".into(),
            ..make_test_reader()
        };
        let mut deltas = HashMap::new();
        deltas.insert("layerName = A".into(), 100);
        deltas.insert("layerName = B".into(), 201); // > 2× of 100
        let (layer, delta) = reader.select_layer(&deltas);
        assert_eq!(layer, "layerName = B");
        assert_eq!(delta, 201);
    }

    #[test]
    fn select_layer_prefers_blast() {
        let reader = make_test_reader();
        let mut deltas = HashMap::new();
        deltas.insert("layerName = Normal#1".into(), 100);
        deltas.insert("layerName = SurfaceView[BLAST]#1".into(), 100);
        let (layer, _) = reader.select_layer(&deltas);
        assert_eq!(layer, "layerName = SurfaceView[BLAST]#1");
    }

    fn make_test_reader() -> FpsReader {
        FpsReader {
            foreground_pkg: Arc::new(Mutex::new(String::new())),
            prev_counts: HashMap::new(),
            prev_time: None,
            prev_focus_pkg: String::new(),
            tracked_layer: String::new(),
            zero_streak: 0,
            last_frame_ns: 0,
            jank_state: JankState::default(),
            consecutive_errors: 0,
            last_enable_time: None,
            jank_count: 0,
            big_jank_count: 0,
        }
    }
}
